File: apple/TriageCompanion/HTTPClient.swift
import Foundation

struct HTTPClient {
    func data(for request: URLRequest, serviceName: String) async throws -> (Data, HTTPURLResponse) {
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse else {
                throw TriageCompanionError.message("\(serviceName) returned a non-HTTP response.")
            }
            guard (200..<300).contains(httpResponse.statusCode) else {
                throw TriageCompanionError.message("\(serviceName) API error \(httpResponse.statusCode): \(errorPreview(from: data))")
            }

            return (data, httpResponse)
        } catch let error as TriageCompanionError {
            throw error
        } catch {
            throw TriageCompanionError.message("Could not load \(serviceName): \(error.localizedDescription)")
        }
    }

    func decoded<T: Decodable>(
        _ type: T.Type,
        from request: URLRequest,
        serviceName: String,
        decoder: JSONDecoder = JSONDecoder()
    ) async throws -> (T, HTTPURLResponse) {
        let (data, response) = try await data(for: request, serviceName: serviceName)
        do {
            return (try decoder.decode(type, from: data), response)
        } catch {
            throw TriageCompanionError.message("\(serviceName) response could not be decoded: \(error.localizedDescription)")
        }
    }

    func jsonObject(from request: URLRequest, serviceName: String) async throws -> ([String: Any], HTTPURLResponse) {
        let (data, response) = try await data(for: request, serviceName: serviceName)
        let value = try JSONSerialization.jsonObject(with: data)
        guard let object = value as? [String: Any] else {
            throw TriageCompanionError.message("\(serviceName) response must be a JSON object.")
        }

        return (object, response)
    }

    func requestWithoutBody(_ request: URLRequest, serviceName: String) async throws {
        _ = try await data(for: request, serviceName: serviceName)
    }

    private func errorPreview(from data: Data) -> String {
        let text = String(data: data, encoding: .utf8) ?? "response body was not UTF-8"
        let normalized = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if normalized.isEmpty {
            return "response body was empty."
        }

        return String(normalized.prefix(500))
    }
}

func request(
    url: URL,
    method: String = "GET",
    headers: [String: String],
    cachePolicy: URLRequest.CachePolicy = .reloadIgnoringLocalCacheData
) -> URLRequest {
    var request = URLRequest(url: url, cachePolicy: cachePolicy)
    request.httpMethod = method
    for (key, value) in headers {
        request.setValue(value, forHTTPHeaderField: key)
    }

    return request
}

func urlWithQuery(_ baseURL: URL, queryItems: [URLQueryItem]) -> URL {
    var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
    components.queryItems = queryItems
    return components.url!
}

func nextURL(from linkHeader: String?) throws -> URL? {
    guard let linkHeader else {
        return nil
    }

    for entry in linkHeader.split(separator: ",") {
        let pieces = entry.split(separator: ";").map { $0.trimmingCharacters(in: .whitespaces) }
        guard pieces.contains("rel=\"next\"") else {
            continue
        }
        guard let first = pieces.first, first.hasPrefix("<"), first.hasSuffix(">") else {
            throw TriageCompanionError.message("Pagination link header was malformed.")
        }

        let urlText = String(first.dropFirst().dropLast())
        guard let url = URL(string: urlText) else {
            throw TriageCompanionError.message("Pagination link header included an invalid URL.")
        }

        return url
    }

    return nil
}

func encodedPathComponent(_ value: String) -> String {
    var allowed = CharacterSet.urlPathAllowed
    allowed.remove(charactersIn: "/")
    return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
}

File: android/app/src/main/kotlin/com/triagecompanion/app/HttpClient.kt
package com.triagecompanion.app

import java.io.IOException
import java.net.URLEncoder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.json.JSONTokener

class HttpResult(val body: String, val linkHeader: String?)

class HttpClient {
    private val client = OkHttpClient()

    suspend fun data(
        url: String,
        method: String = "GET",
        headers: Map<String, String>,
        serviceName: String,
    ): HttpResult = withContext(Dispatchers.IO) {
        val builder = try {
            Request.Builder().url(url)
        } catch (_: IllegalArgumentException) {
            throw TriageCompanionException("Could not load $serviceName: request URL was invalid.")
        }
        builder.method(method, if (method == "GET") null else ByteArray(0).toRequestBody(null))
        for ((name, value) in headers) {
            builder.header(name, value)
        }

        val response = try {
            client.newCall(builder.build()).execute()
        } catch (error: IOException) {
            throw TriageCompanionException("Could not load $serviceName: ${error.message ?: error.javaClass.simpleName}")
        }
        response.use {
            val body = it.body?.string() ?: ""
            if (it.code !in 200..299) {
                throw TriageCompanionException("$serviceName API error ${it.code}: ${errorPreview(body)}")
            }

            HttpResult(body, it.header("Link"))
        }
    }

    suspend fun jsonObject(
        url: String,
        headers: Map<String, String>,
        serviceName: String,
    ): Pair<JSONObject, HttpResult> {
        val result = data(url, headers = headers, serviceName = serviceName)
        val value = try {
            JSONTokener(result.body).nextValue()
        } catch (error: JSONException) {
            throw TriageCompanionException("$serviceName response could not be decoded: ${error.message}")
        }
        if (value !is JSONObject) {
            throw TriageCompanionException("$serviceName response must be a JSON object.")
        }

        return value to result
    }

    suspend fun jsonArray(
        url: String,
        headers: Map<String, String>,
        serviceName: String,
    ): Pair<JSONArray, HttpResult> {
        val result = data(url, headers = headers, serviceName = serviceName)
        val value = try {
            JSONTokener(result.body).nextValue()
        } catch (error: JSONException) {
            throw TriageCompanionException("$serviceName response could not be decoded: ${error.message}")
        }
        if (value !is JSONArray) {
            throw TriageCompanionException("$serviceName response must be a JSON array.")
        }

        return value to result
    }

    suspend fun requestWithoutBody(
        url: String,
        method: String,
        headers: Map<String, String>,
        serviceName: String,
    ) {
        data(url, method, headers, serviceName)
    }

    private fun errorPreview(body: String): String {
        val normalized = body.trim()
        if (normalized.isEmpty()) {
            return "response body was empty."
        }

        return normalized.take(500)
    }
}

fun urlWithQuery(baseURL: String, queryItems: List<Pair<String, String>>): String {
    if (queryItems.isEmpty()) {
        return baseURL
    }

    val query = queryItems.joinToString("&") { (name, value) ->
        "${encodedQueryComponent(name)}=${encodedQueryComponent(value)}"
    }
    return "$baseURL?$query"
}

fun nextURL(linkHeader: String?): String? {
    if (linkHeader == null) {
        return null
    }

    for (entry in linkHeader.split(",")) {
        val pieces = entry.split(";").map { it.trim() }
        if (!pieces.contains("rel=\"next\"")) {
            continue
        }
        val first = pieces.firstOrNull()
        if (first == null || !first.startsWith("<") || !first.endsWith(">")) {
            throw TriageCompanionException("Pagination link header was malformed.")
        }

        return first.substring(1, first.length - 1)
    }

    return null
}

fun encodedPathComponent(value: String): String =
    URLEncoder.encode(value, "UTF-8").replace("+", "%20")

fun encodedQueryComponent(value: String): String =
    URLEncoder.encode(value, "UTF-8").replace("+", "%20")

fun queryPairs(uri: java.net.URI): List<Pair<String, String?>> {
    val raw = uri.rawQuery ?: return emptyList()
    if (raw.isEmpty()) {
        return emptyList()
    }

    return raw.split("&").map { entry ->
        val separator = entry.indexOf('=')
        if (separator < 0) {
            (percentDecoded(entry) ?: entry) to null
        } else {
            val name = entry.substring(0, separator)
            val value = entry.substring(separator + 1)
            (percentDecoded(name) ?: name) to (percentDecoded(value) ?: value)
        }
    }
}

fun sortedQueryString(uri: java.net.URI, excluding: Set<String>): String =
    queryPairs(uri)
        .filter { it.first !in excluding }
        .sortedWith(compareBy({ it.first }, { it.second ?: "" }))
        .joinToString("&") { (name, value) ->
            if (value != null) {
                "${encodedQueryComponent(name)}=${encodedQueryComponent(value)}"
            } else {
                encodedQueryComponent(name)
            }
        }

fun queryValue(uri: java.net.URI, name: String): String? =
    queryPairs(uri).firstOrNull { it.first == name }?.second

fun percentDecoded(value: String): String? {
    val bytes = java.io.ByteArrayOutputStream()
    val literal = StringBuilder()
    var index = 0
    while (index < value.length) {
        val character = value[index]
        if (character == '%') {
            if (index + 2 >= value.length) {
                return null
            }
            val decoded = value.substring(index + 1, index + 3).toIntOrNull(16) ?: return null
            bytes.write(literal.toString().toByteArray(Charsets.UTF_8))
            literal.setLength(0)
            bytes.write(decoded)
            index += 3
        } else {
            literal.append(character)
            index += 1
        }
    }
    bytes.write(literal.toString().toByteArray(Charsets.UTF_8))

    return String(bytes.toByteArray(), Charsets.UTF_8)
}
